module/zfs/spa.c

```diff
@@ -9591,6 +9604,153 @@ spa_sync_allpools(void)
 	mutex_exit(&spa_namespace_lock);
 }
 
+static void spa_sync_tq_assign(void *arg);
+
+typedef struct sync_tq_arg {
+	kthread_t	*sta_thread;
+	kcondvar_t	sta_cv;
+	kmutex_t 	sta_lock;
+	int		sta_ready;
+} sync_tq_arg_t;
+
+/* unfortunately, taskqueues do not provide per-thread private data */
+static void
+spa_sync_tq_assign(void *arg)
+{
+	sync_tq_arg_t *sta = arg;
+
+	mutex_enter(&sta->sta_lock);
+	sta->sta_thread = curthread;
```

Comment:
At least on FreeBSD and Illumos taskq is a (wrapper around) kernel KPI, and it would require OS change to implement taskq_threads().

Reply:
Then perhaps we could implement a limited taskq tsd interface which could be used.  It looks like the FreeBSD taskq code already uses `tsd_get()/set()` and this interface is available on Illumos as well.  My understanding is both of these implementations are quite efficient.  Then generic tsd implementation added for Linux might be good enough (although I doubt it), if not we could always reserve stack space `taskq_thread()` for this.

Reply:
@behlendorf  would that still need a way to ensure that each thread in the taskq pool runs once (so we can get its curthread)?

Reply:
> This worries me because the taskq API makes absolutely no promises how the tasks will run on the thread pool. This only happens to work out because all the implementations (I'm aware of) split the tasks evenly over the threads.

AFAICS tho not specifically stated, it seems the API implicitly requires such splitting in any implementation, if it is to make efficient use of the pool threads and avoid unneeded latencies. So a reliable use case would have to perform the following (which this patch does):

 - The number of taskq_dispatches equals the number of threads in the pool.
 - The task_func_t does not return (blocks) until the dispatches to all threads have been issued (hence the cv_ syncing)
 - The taskq_create, taskq_dispatches, and cv_ syncs are all done from the same thread and in that order.
 - Use TASKQ_PREPOPULATE, !TASKQ_DYNAMIC, and TQ_FRONT.

Perhaps it would be good to encapsulate the mechanism and put it into the SPL (e.g., taskq_create_synced(...) or such).  BTW, this does not assume/depend on any actual thread execution order.  Thoughts?

Reply:
> Perhaps it would be good to encapsulate the mechanism and put it into the SPL.

If we're going to depend on this behavior extending the SPL taskq API to provide it seems like a reasonable way handle this.  I'd be fine with adding a `taskq_create_synced()` function which also returns the `kthread_t *` for the newly created thread pool.  This would at least allow each implementation to handle this in both an optimal and guaranteed correct way.

Reply:
I might be oversimplifying, but I feel like this is not so far from `TASKQ_PREPOPULATE|TASKQ_DYNAMIC`, with a rough notion of the workers being partitioned?

I don't think I'd push for major changes, since I feel like the spa+taskq setup is already creaking at the seams and this PR is both pushing it further but also trying to fix up some of its shortcomings, so I'd like to see the step forward. But I might have at least hidden the thread array inside `taskq_t` so it can be thought of as a single unit: 51e96eb8

Reply:
updated to implement taskq_create_synced() and taskq_destroy_synced per above.